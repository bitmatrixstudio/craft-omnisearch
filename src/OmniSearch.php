<?php
/*
 * @copyright Copyright (c) 2021 Bitmatrix Studio
 * @license https://craftcms.github.io/license/
 */

namespace pohnean\omnisearch;

use Craft;
use craft\base\Plugin;
use craft\controllers\TemplatesController;
use craft\elements\db\ElementQuery;
use craft\elements\db\EntryQuery;
use craft\events\DefineBehaviorsEvent;
use craft\helpers\Json;
use craft\web\Controller;
use craft\web\View;
use pohnean\omnisearch\assetbundles\omnisearch\OmniSearchAsset;
use pohnean\omnisearch\behaviors\OmniSearchFilterBehavior;
use pohnean\omnisearch\models\Settings;
use pohnean\omnisearch\services\OmniSearchService as OmniSearchServiceService;
use yii\base\Event;

/**
 * Class OmniSearch
 *
 * @author    Bitmatrix Studio
 * @package   OmniSearch
 * @since     1.0.0
 *
 * @property  OmniSearchServiceService $omniSearchService
 */
class OmniSearch extends Plugin
{
	const DATATYPE_TEXT = 'text';
	const DATATYPE_NUMBER = 'number';
	const DATATYPE_DATE = 'date';
	const DATATYPE_BOOLEAN = 'boolean';
	const DATATYPE_LIST = 'list';
	const DATATYPE_MATRIX = 'matrix';

	// Static Properties
	// =========================================================================

	/**
	 * @var OmniSearch
	 */
	public static $plugin;

	// Public Properties
	// =========================================================================

	/**
	 * @var string
	 */
	public $schemaVersion = '1.0.0';

	// Public Methods
	// =========================================================================

	/**
	 * @inheritdoc
	 */
	public function init()
	{
		parent::init();
		self::$plugin = $this;

		if (Craft::$app->getRequest()->isCpRequest) {
			Craft::$app->getView()->registerAssetBundle(OmniSearchAsset::class);
		}

		Event::on(ElementQuery::class, EntryQuery::EVENT_DEFINE_BEHAVIORS, function (DefineBehaviorsEvent $event) {
			/** @var ElementQuery $sender */
			$sender = $event->sender;
			$sender->attachBehavior('omnisearch', new OmniSearchFilterBehavior());
		});
	}

	// Protected Methods
	// =========================================================================

	/**
	 * @inheritdoc
	 */
	protected function createSettingsModel()
	{
		return new Settings();
	}

	/**
	 * @inheritdoc
	 */
	protected function settingsHtml(): string
	{
		return Craft::$app->view->renderTemplate(
			'omnisearch/settings',
			[
				'settings' => $this->getSettings()
			]
		);
	}
}
