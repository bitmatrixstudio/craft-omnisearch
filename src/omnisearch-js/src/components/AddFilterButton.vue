<template>
  <div class="omnisearch__add-filter">
    <button type="button"
            class="btn small icon omnisearch__add-filter-btn"
            @click="toggleMenu"
            ref="button"
    >
      <template v-if="selectedField != null">
        <strong>{{ buttonText }}</strong>{{ operatorText }}
      </template>
      <template v-else>{{ buttonText }}</template>
    </button>
    <div v-if="showFieldMenu"
         class="menu omnisearch__filter-panel omnisearch__choose-fields"
         ref="filterPanel"
         data-test="filterPanel"
    >
      <template v-if="selectedField == null">
        <!-- Step 1: choose field from list -->
        <div class="omnisearch__field-list-search">
          <div class="flex-grow texticon search icon">
            <input class="text"
                   type="text"
                   v-model="keyword"
                   placeholder="Search attributes..."
                   ref="searchInput"
                   data-test="fieldSearchInput"
            />
          </div>
        </div>
        <div class="omnisearch__filter-panel-body" data-test="fieldList">
          <div class="omnisearch__list-item"
               data-test="fieldListItem"
               v-for="(field, index) in fieldList"
               :key="index"
               @click="setSelectedField(field)">
            {{ field.name }}
          </div>
        </div>
      </template>
      <template v-else-if="selectedFilterMethod == null">
        <!-- Step 2: choose filter method -->
        <div class="omnisearch__filter-panel-body" data-test="filterMethodList">
          <div class="omnisearch__list-item"
               data-test="filterMethodListItem"
               v-for="filterMethod in filterMethods"
               :key="filterMethod.operator"
               @click="selectFilterMethod(filterMethod)"
          >
            {{ filterMethod.label }}
          </div>
        </div>
      </template>
      <template v-else>
        <!-- Step 3: choose value -->
        <div class="omnisearch__filter-panel-body" data-test="compareValue">
          <input
            ref="compareValueTextInput"
            data-test="compareValueTextInput"
            class="text"
            type="text"
            v-model="compareValue"
            @keydown.enter="applyFilter"
          />
        </div>
        <div class="omnisearch__filter-panel-footer">
          <button
            class="btn fullwidth"
            :class="{ disabled: compareValue == null }"
            type="button"
            :disabled="compareValue == null"
            @click="applyFilter"
            data-test="applyFilterBtn">
            Apply Filter
          </button>
        </div>
      </template>
    </div>
  </div>
</template>

<script>
import { createPopper } from '@popperjs/core';
import sortBy from 'lodash/sortBy';

export default {
  name: 'AddFilterButton',
  props: {
    fields: {
      type: Array,
      required: true,
      default: () => ([]),
    },
  },
  data() {
    return {
      keyword: '',
      selectedField: null,
      selectedFilterMethod: null,
      compareValue: null,
      showFieldMenu: false,
    };
  },
  computed: {
    buttonText() {
      if (this.showFieldMenu) {
        return this.selectedField ? this.selectedField.name : 'Choose Field';
      }

      return '+ Add Filter';
    },
    operatorText() {
      if (this.selectedFilterMethod != null) {
        return ` ${this.selectedFilterMethod.label}`;
      }

      return '';
    },
    fieldList() {
      const filteredFields = this.fields.filter(
        (field) => field.name.toLowerCase().includes(this.keyword.toLowerCase()),
      );

      return sortBy(filteredFields, 'name');
    },
    filterMethods() {
      return [
        { operator: 'contain', label: 'contains' },
        { operator: 'not_contain', label: 'does not contains' },
        { operator: 'equal', label: 'equals' },
        { operator: 'not_equal', label: 'not equal to' },
        { operator: 'starts_with', label: 'starts with' },
        { operator: 'is_present', label: 'is present', requiresValue: false },
        { operator: 'is_not_present', label: 'is not present', requiresValue: false },
      ];
    },
  },
  watch: {
    showFieldMenu(show) {
      if (show) {
        this.reset();

        this.$nextTick(() => {
          this.popper = createPopper(this.$refs.button, this.$refs.filterPanel, {
            placement: 'bottom-start',
          });

          this.$refs.searchInput.focus();
        });
      } else if (!show && this.popper) {
        this.popper.destroy();
      }
    },
  },
  methods: {
    reset() {
      this.selectedField = null;
      this.selectedFilterMethod = null;
      this.compareValue = null;
    },
    toggleMenu() {
      this.showFieldMenu = !this.showFieldMenu;
    },
    setSelectedField(field) {
      this.selectedField = field;
    },
    selectFilterMethod(method) {
      const { requiresValue = true } = method;

      this.selectedFilterMethod = method;

      if (!requiresValue) {
        this.applyFilter();
      } else {
        this.$nextTick(() => {
          if (this.$refs.compareValueTextInput != null) {
            this.$refs.compareValueTextInput.focus();
          }
        });
      }
    },
    applyFilter() {
      const { requiresValue = true } = this.selectedFilterMethod;

      if (!requiresValue || (this.compareValue !== null && this.compareValue !== '')) {
        this.$emit('add-filter', {
          field: this.selectedField.handle,
          operator: this.selectedFilterMethod.operator,
          value: this.compareValue,
        });

        this.showFieldMenu = false;
        this.reset();
      }
    },
  },
  beforeDestroy() {
    if (this.popper) {
      this.popper.destroy();
    }
  },
};
</script>

<style lang="scss">
  .omnisearch__add-filter {
    display: inline-block;
  }
</style>
