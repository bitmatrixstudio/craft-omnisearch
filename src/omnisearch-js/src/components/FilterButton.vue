<template>
  <div class="omnisearch__filter" :data-testid="testId">
    <div
      ref="button"
      data-testid="filter-button"
      class="btn small icon"
      @click="onFilterButtonClick"
    >
      <span class="omnisearch__filter-text">
        <template v-if="selectedField != null"><strong>{{
            selectedField.name
          }}</strong>{{ operatorText }} {{ valueText }}</template>
        <template v-else-if="showFilterPanel">Choose Field</template>
        <template v-else>+ Add Filter</template>
      </span>
      <button
        v-if="!isNewFilter"
        type="button"
        class="omnisearch__remove-filter-btn"
        data-testid="remove-filter-button"
        @click.stop="removeFilter">
        &times;
      </button>
    </div>
    <filter-panel
      v-if="showFilterPanel"
      :fields="fields"
      :selected-field.sync="selectedField"
      :selected-filter-method.sync="selectedFilterMethod"
      :compare-value.sync="compareValue"
      ref="filterPanel"
      @apply="applyFilter"
    />
  </div>
</template>

<script>
import { createPopper } from '@popperjs/core';
import operators from '../operators';
import { formatValues } from '../formatters';
import FilterPanel from './FilterPanel.vue';

export default {
  name: 'FilterButton',
  components: {
    FilterPanel,
  },
  props: {
    fields: {
      type: Array,
      required: true,
      default: () => ([]),
    },
    filter: {
      type: Object,
    },
    testId: {
      type: String,
    },
  },
  data() {
    return {
      showFilterPanel: false,
      selectedField: null,
      selectedFilterMethod: null,
      compareValue: null,
    };
  },
  computed: {
    isNewFilter() {
      return this.filter == null;
    },
    operatorText() {
      return this.selectedFilterMethod != null ? ` ${this.selectedFilterMethod.label}` : '';
    },
    valueText() {
      const { selectedField, compareValue, selectedFilterMethod } = this;

      if (selectedField == null || selectedFilterMethod == null || !this.hasValue) {
        return '';
      }

      return formatValues(selectedField, selectedFilterMethod.operator, compareValue);
    },
    hasValue() {
      const { compareValue } = this;

      const isNonEmptyScalar = !Array.isArray(compareValue)
        && (compareValue !== null && compareValue !== '');
      const isNonEmptyArray = Array.isArray(compareValue) && compareValue.length > 0;

      return isNonEmptyScalar || isNonEmptyArray;
    },
  },
  watch: {
    showFilterPanel(show) {
      if (show) {
        this.$nextTick(() => {
          this.popper = createPopper(this.$refs.button, this.$refs.filterPanel.$el, {
            placement: 'bottom-start',
          });
        });
      } else if (!show && this.popper) {
        this.popper.destroy();
      }
    },
  },
  methods: {
    onFilterButtonClick() {
      const { showFilterPanel } = this;

      if (!showFilterPanel) {
        this.showFilterPanel = true;
      } else {
        this.closePanel(true);
      }
    },
    reset() {
      const { filter, isNewFilter } = this;

      if (isNewFilter) {
        this.selectedField = null;
        this.selectedFilterMethod = null;
        this.compareValue = null;
      } else {
        this.selectedField = this.fields.find((field) => field.handle === filter.field);
        this.selectedFilterMethod = operators.find(
          (operator) => operator.operator === filter.operator,
        );
        this.compareValue = filter.value;
      }
    },
    applyFilter() {
      const { requiresValue = true } = this.selectedFilterMethod;

      if (!requiresValue || this.hasValue) {
        this.$emit('apply', {
          field: this.selectedField.handle,
          operator: this.selectedFilterMethod.operator,
          value: this.compareValue,
        });
      }

      this.closePanel(this.isNewFilter);
    },
    removeFilter() {
      this.$emit('remove');
    },
    closePanel(reset = false) {
      this.showFilterPanel = false;

      if (reset) {
        this.reset();
      }
    },
    handleClickOutside(e) {
      if (this.showFilterPanel === false) {
        return;
      }

      if (this.$refs.filterPanel
        && !this.$refs.filterPanel.$el.contains(e.target)
        && this.$refs.button
        && !this.$refs.button.contains(e.target)) {
        this.closePanel(true);
      }
    },
  },
  created() {
    this.reset();
  },
  mounted() {
    document.addEventListener('click', this.handleClickOutside, false);
  },
  beforeDestroy() {
    if (this.popper) {
      this.popper.destroy();
    }

    document.removeEventListener('click', this.handleClickOutside);
  },
};
</script>

<style lang="scss">
.omnisearch__add-filter {
  display: inline-block;
}

.omnisearch__filter-panel-body > input {
  width: 100%;
}

.omnisearch__filter-method-dropdown {
  display: flex;
  justify-content: space-between;
  width: 100%;
  margin-bottom: 0.5rem;

  &:after {
    display: block;
    content: '.';
    font-size: 0;
    width: 5px;
    height: 5px;
    border: solid #596673;
    border-width: 0 2px 2px 0;
    -webkit-transform: rotate(
        45deg);
    -o-transform: rotate(45deg);
    transform: rotate(45deg);
    position: relative;
  }
}

.omnisearch__filter {
  margin-right: 0.5em;

  > .btn {
    display: flex;
    align-items: center;
  }
}

.omnisearch__remove-filter-btn {
  border: none;
  border-radius: 50%;
  padding: 0;
  width: 16px;
  line-height: 1rem;
  margin-left: 0.25em;
  font-size: 16px;
  background-color: transparent;
  cursor: pointer;

  &:hover {
    color: #F00;
  }
}
</style>
