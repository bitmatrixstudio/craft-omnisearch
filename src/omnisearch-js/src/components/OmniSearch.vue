<!--
  - @copyright Copyright (c) 2021 Bitmatrix Studio
  - @license https://craftcms.github.io/license/
  -->

<template>
  <div class="omnisearch">
    <template v-if="fields.length > 0 && activeFilters.length > 0">
      <filter-button
        v-for="(filter, index) in activeFilters"
        :fields="fields"
        :filter="filter"
        :test-id="`active-filter-${index}`"
        :key="`filter-${index}`"
        @apply="updateFilter(filter, $event)"
        @remove="removeFilter(index)"
      />
    </template>
    <filter-button
      test-id="add-filter"
      :fields="fields"
      key="add-filter"
      @apply="addFilter"
    />
  </div>
</template>

<script>
import FilterButton from './FilterButton.vue';
import DATATYPES from '../datatypes';

export default {
  name: 'OmniSearch',
  components: { FilterButton },
  props: {
    fields: {
      type: Array,
      required: true,
      default: () => ([]),
    },
    initialFilters: {
      type: Array,
      default: () => ([]),
    },
  },
  data() {
    return {
      activeFilters: this.processFilters(this.initialFilters),
    };
  },
  computed: {
    fieldMap() {
      return this.fields.reduce((fieldMap, field) => {
        Object.assign(fieldMap, {
          [field.handle]: field,
        });

        return fieldMap;
      }, {});
    },
  },
  watch: {
    initialFilters(newValue) {
      this.activeFilters = this.processFilters(newValue);
    },
    activeFilters: {
      deep: true,
      handler() {
        this.updateFilterCriteria();
      },
    },
  },
  methods: {
    reset() {
      this.activeFilters = [];
    },
    getField(handle) {
      return this.fieldMap[handle] != null ? this.fieldMap[handle] : null;
    },
    addFilter(filter) {
      this.activeFilters.push(filter);
    },
    updateFilter(filter, data) {
      Object.assign(filter, data);
    },
    removeFilter(index) {
      this.activeFilters.splice(index, 1);
    },
    updateFilterCriteria() {
      this.$emit('change', this.activeFilters);
    },
    processFilters(filters) {
      const { fields } = this;

      return filters.map((filter) => {
        const field = fields.find((item) => item.handle === filter.field);
        if (!field) {
          return filter;
        }

        return {
          ...filter,
          value: this.processValue(field.dataType, filter.value),
        };
      });
    },
    processValue(dataType, value) {
      switch (dataType) {
        case DATATYPES.BOOLEAN: {
          return value === 1 || value === '1' || value === 'true';
        }

        case DATATYPES.NUMBER: {
          return Number(value);
        }

        default: {
          return value;
        }
      }
    },
  },
};
</script>

<style lang="scss">
.omnisearch {
  position: relative;
  margin-top: -1em;
  margin-bottom: 1em;
  display: inline-flex;
  flex-wrap: wrap;

  .btn:not(.small) {
    font-size: 1em;
  }

  .omnisearch__filter {
    display: flex;
    margin-bottom: 0.5em;
  }

  .omnisearch__filter-text {
    display: block;
    max-width: 20rem;
    overflow: hidden;
    text-overflow: ellipsis;
  }

  .omnisearch__filter-panel {
    display: block;
    padding: 0 !important;
    min-width: 10rem;
  }

  .omnisearch__field-list-search {
    padding: 0.5rem;
  }

  .omnisearch__field-list-search + .omnisearch__filter-panel-body {
    padding-top: 0;
  }

  .omnisearch__filter-panel-body {
    max-height: 21rem;
    overflow-y: auto;
  }

  .omnisearch__filter-panel-body,
  .omnisearch__filter-panel-footer {
    padding: 0.5rem;
  }

  .omnisearch__list-item {
    cursor: pointer;
    padding: 10px 14px;
    margin: 0 -0.5rem;

    &:hover {
      background-color: #f3f7fc;
    }

    &.omnisearch__list-item-group {
      cursor: default;
      font-weight: bold;

      &:hover {
        background-color: transparent;
      }
    }
  }

  .omnisearch__list-item-group-list {
    .omnisearch__list-item {
      padding-left: 24px;
    }
  }
}
</style>
