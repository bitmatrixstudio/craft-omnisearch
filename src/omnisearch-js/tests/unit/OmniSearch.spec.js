import { shallowMount } from '@vue/test-utils';
import OmniSearch from '../../src/components/OmniSearch.vue';
import AddFilterButton from '../../src/components/AddFilterButton.vue';
import FilterButton from '../../src/components/FilterButton.vue';

describe('OmniSearch.vue', () => {
  let wrapper;
  let mockUpdateElementsFn;

  beforeEach(() => {
    mockUpdateElementsFn = jest.fn();

    window.Craft = {
      elementIndex: {
        settings: {
          context: 'index',
          storageKey: 'elementindex.craft\\elements\\Entry',
          criteria: {
            enabledForSite: null,
          },
        },
        updateElements: mockUpdateElementsFn,
      },
    };
  });

  describe('without initial filters', () => {
    beforeEach(() => {
      wrapper = shallowMount(OmniSearch, {
        propsData: {
          fields: [
            {
              fieldName: 'Title',
            },
            {
              fieldName: 'Post Date',
            },
          ],
        },
      });
    });

    it('renders correctly with no active filters', () => {
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.classes()).toContain('omnisearch');
      expect(wrapper.find('.omnisearch__active-filters').exists()).toBe(false);
      expect(wrapper.findAll(AddFilterButton).length).toBe(1);
    });
  });

  describe('with initial filters', () => {
    beforeEach(() => {
      wrapper = shallowMount(OmniSearch, {
        propsData: {
          fields: [
            {
              fieldName: 'Title',
            },
            {
              fieldName: 'Post Date',
            },
          ],
          initialFilters: [
            {
              field: { fieldName: 'Title' },
              operator: 'is_present',
            },
          ],
        },
      });
    });

    it('renders correctly with initial active filters', () => {
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.findAll('.omnisearch__active-filters').length).toEqual(1);
    });

    it('should add to activeFilters array and performSearch when addFilter event is emitted',
      async () => {
        wrapper.find(AddFilterButton).vm.$emit('add-filter', {
          field: { fieldName: 'Title' },
          operator: 'equals',
        });

        expect(wrapper.vm.activeFilters.length).toBe(2);
        await wrapper.vm.$nextTick();
        expect(window.Craft.elementIndex.settings.criteria.omnisearchFilters.length).toBe(2);
        expect(mockUpdateElementsFn).toHaveBeenCalledTimes(1);
      });

    it('should remove from activeFilters when removeFilter event is emitted', async () => {
      wrapper.find(FilterButton).vm.$emit('remove-filter', 0);

      expect(wrapper.vm.activeFilters.length).toBe(0);
      await wrapper.vm.$nextTick();
      expect(window.Craft.elementIndex.settings.criteria.omnisearchFilters.length).toBe(0);
      expect(mockUpdateElementsFn).toHaveBeenCalledTimes(1);
    });
  });

  // it('should perform a search when activeFilters change', () => {
  //
  // });
});
