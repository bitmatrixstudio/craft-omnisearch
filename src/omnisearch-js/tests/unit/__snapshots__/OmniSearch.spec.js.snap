// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OmniSearch.vue with initial filters renders correctly with initial active filters 1`] = `
<div
  class="omnisearch"
>
  <div
    class="omnisearch__active-filters"
  >
    <filter-button-stub
      fieldname="Title"
      operator="is_present"
    />
  </div>
   
  <add-filter-button-stub
    fields="[object Object],[object Object]"
  />
</div>
`;

exports[`OmniSearch.vue without initial filters renders correctly with no active filters 1`] = `
<div
  class="omnisearch"
>
  <!---->
   
  <add-filter-button-stub
    fields="[object Object],[object Object]"
  />
</div>
`;
